src/rocksdb_options.rs

```diff
@@ -1246,11 +1245,32 @@ impl ColumnFamilyOptions {
                 Ok(s) => s,
                 Err(e) => return Err(format!("failed to convert to cstring: {:?}", e)),
             };
-            self.filter = Some(new_compaction_filter(c_name, ignore_snapshots, filter)?);
-            crocksdb_ffi::crocksdb_options_set_compaction_filter(
-                self.inner,
-                self.filter.as_ref().unwrap().inner,
-            );
+            let filter = new_compaction_filter(c_name, filter);
+            crocksdb_ffi::crocksdb_options_set_compaction_filter(self.inner, filter.inner);
+            self.filter = Some(filter);
+            Ok(())
+        }
+    }
+
+    /// Set compaction filter factory.
+    ///
+    /// See also `CompactionFilterFactory`.
+    pub fn set_compaction_filter_factory<S>(
+        &mut self,
+        name: S,
+        factory: Box<dyn CompactionFilterFactory>,
+    ) -> Result<(), String>
+    where
+        S: Into<Vec<u8>>,
+    {
+        let c_name = match CString::new(name) {
+            Ok(s) => s,
+            Err(e) => return Err(format!("failed to convert to cstring: {:?}", e)),
+        };
+        unsafe {
+            let factory = new_compaction_filter_factory(c_name, factory)?;
+            crocksdb_ffi::crocksdb_options_set_compaction_filter_factory(self.inner, factory.inner);
+            std::mem::forget(factory); // Deconstructor will be called after `self` is dropped.
```

Comment:
Just curious, why `mem::forget` is still needed? The factory is moved into `new_compaction_filter_factory`, I think?

Reply:
`factory.inner` is moved into `options` as a shared ptr. But `factory` is a `struct`, so deconstructor will be called if we don't forget it.